### PR TITLE
fix: Handle side menu navigation correctly

### LINKED ISSUE #5123: Public Event Page Menu: Moving to Sponsors, Organizers etc. requires two clicks
To get to sections on the same page that are listed on the menu requires two clicks, e.g. sponsors or organizers.

Compare https://eventyay.com/e/16fa59c7

Screenshot from 2020-09-23 00-38-17

### PR DESCRIPTION
**Pull request by daretobedifferent18**
Fixes #5123 

#### Short description of what this resolves:
Require two clicks for moving to same section eg. sponsors, tickets etc.

#### Changes proposed in this pull request:

- Since the problem what I found is that eventListeners were no longer there for each button until an unless we click on any button. 
- EventListeners need to be with each button before, so I have resolved this by adding life cycle Hook
- Now it requires only one click as normal to move to diff sections of the page.
- Going to sponsors, getting here etc from speaker, call for speaker etc pages were not exactly taking to the section of page so fixed this.
- Also clicking sponsors, organizers etc from diff page gets to public.index page but the info button remain active so fixed this.

### TODO

- [x] Single click and go to section on same page
- [x] Target to specific section from diff page.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by daretobedifferent18** [comment]
@iamareebjamal Please see to the changes I have made in it and let me know anything to do.

**Comment by iamareebjamal** [comment]
Go to https://open-event-frontend-i18x7pe1p.vercel.app/e/1f720482/speakers
Click on sponsors
Not working

**Comment by daretobedifferent18** [comment]
> https://open-event-frontend-i18x7pe1p.vercel.app/e/1f720482/speakers
Now I have to work on when going from different page to diff page section but being on same page its working

**Comment by iamareebjamal** [comment]
Yes

**Comment by daretobedifferent18** [comment]
@iamareebjamal can we use ```this.transitionToRoute``` in component js file?

**Comment by iamareebjamal** [comment]
Yes

**Comment by daretobedifferent18** [comment]
@iamareebjamal made the changes. Now getting to section from diff page is working properly.

**Comment by daretobedifferent18** [comment]
@iamareebjamal I mean i got the way how it will be working and its working what was said but the problem that will be arising here is that I have used setTimeout funtion for 2 sec but this won't work always bcz what if page took more than 2 sec so I want to know is there a way so that function gets executed only when the page is completely loaded?

**Comment by daretobedifferent18** [comment]
@daretobedifferent18 https://open-event-frontend-mq4nd06op.vercel.app/e/1f720482/speakers 
and clicking on sponsors is working.

**Comment by iamareebjamal** [comment]
Fix the build and indentation. Cannot review till then

File: app/controllers/public.js [hunk 1]
```diff
@@ -37,4 +37,14 @@ export default class PublicController extends Controller {
   toggleMenu() {
     this.toggleProperty('isMenuOpen');
   }
+
+  @action
+  goToLink(section) {
+      this.transitionToRoute('public.index');
+      setTimeout(function(){
```
Comment (iamareebjamal) [inline app/controllers/public.js:44]: This will most likely fail if the transition is not completed within the timeout

Comment (daretobedifferent18) [reply app/controllers/public.js]: Is their other way so that function gets executed only when the page is loaded?

**Comment by daretobedifferent18** [comment]
@iamareebjamal Is their other way so that function gets executed only when the page is loaded?

File: app/controllers/public.js [hunk 2]
```diff
@@ -37,4 +37,13 @@ export default class PublicController extends Controller {
   toggleMenu() {
     this.toggleProperty('isMenuOpen');
   }
+
+  @action
+  goToLink(section) {
+    this.transitionToRoute('public.index');
+    setTimeout(function() {
+      document.querySelector(`[href='#${section}']`).click();
+    }, 2000);
```
Comment (iamareebjamal) [inline app/controllers/public.js:46]: Too much waiting. You should not depend on random waits and timeouts. See if `transitionToRoute` returns a promise and then await it

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal now what I found is that I can do 
```javascript
...

transitionToRoute('public.index').promise.then(
   document.querySelector(`[href='#${section}']`).click();
)

... 
```

but this is not working

Comment (iamareebjamal) [reply app/controllers/public.js:46]: This is not even valid JS syntax. Of course it won't work

Comment (iamareebjamal) [reply app/controllers/public.js:46]: Make action async and await on transitionToRoute

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal  its not working with this also

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: transitionToRoute promise completes before it finished rendering items

Comment (iamareebjamal) [reply app/controllers/public.js:46]: Then your first solution seems like the best option

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal setTimeout one?

Comment (iamareebjamal) [reply app/controllers/public.js:46]: Yes, that was your first solution

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal OK so should I change the time, or 2sec is ok? and if everything else is fine then I think this is done. Let me know anything to change.

**Comment by daretobedifferent18** [comment]
@iamareebjamal This is the build error that is coming
Screenshot 2020-11-08 031455

File: app/controllers/public.js (hunk 2 above)
Comment (iamareebjamal) [reply app/controllers/public.js:46]: 2 sec was not in your first solution.

> Then your first solution seems like the best option

Read your first solution again

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: So you are talking about my first commit 'Fix Double Click' ?

Comment (iamareebjamal) [reply app/controllers/public.js:46]: I'm talking about first working solution which was setTimeout without delay

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: so sorry But it was always with 2000ms, I never set setTimeout with 0ms.
Screenshot from 2020-11-08 03-38-17

**Comment by iamareebjamal** [comment]
That;s not failing the lint. Look below

File: app/controllers/public.js (hunk 2 above)
Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal First solution was for fixing button for same page 
second one was for fixing from different page which is above 
third was improving the above code and using click() handler.
and in second and third settimeout with 2000ms was used. So pls let me know if I am missing something to look at?

Comment (daretobedifferent18) [reply app/controllers/public.js:46]: @iamareebjamal please let me know

**Comment by daretobedifferent18** [comment]
@iamareebjamal First solution was for fixing button for same page
second one was for fixing from different page which is above
third was improving the above code and using click() handler.
and in second and third settimeout with 2000ms was used. So pls let me know if I am missing something to look at?
please let me know

**Comment by iamareebjamal** [comment]
Fix the build first please

**Comment by daretobedifferent18** [comment]
@iamareebjamal fixed build. Now let me know what to do?

**Comment by daretobedifferent18** [comment]
@iamareebjamal PLease let me know

**Comment by daretobedifferent18** [comment]
@iamareebjamal Please let me know if I have to change anything now bcz build is doing fine. If not then I can take up other issues and start working on it.

**Comment by iamareebjamal** [comment]
I don't understand why, for some people, it is important to get something reviewed/merged to work on other issues? You can still work on other issues. Your work on the PR is complete. I'll review when I get time, I'm busy in other high priority issues right now. This is not the kind of PR which I just look  at the code and merge. I'll have to test it thoroughly and research if there is a better way. Why is that stopping you from working on other stuff? If there is something to change, I'll comment and then you can revisit it. Take a look at other OSS projects, it takes months to get the PRs merged

**Comment by daretobedifferent18** [comment]
> I don't understand why, for some people, it is important to get something reviewed/merged to work on other issues? You can still work on other issues. Your work on the PR is complete. I'll review when I get time, I'm busy in other high priority issues right now. This is not the kind of PR which I just look at the code and merge. I'll have to test it thoroughly and research if there is a better way. Why is that stopping you from working on other stuff? If there is something to change, I'll comment and then you can revisit it. Take a look at other OSS projects, it takes months to get the PRs merged

I thought that once the PR is merged or once the mentor reviewing the PR says that now you can go and take other issue, we can't work on other issues but now I am clear. Sorry for misunderstanding. I'll be taking care of this fact.

**Comment by daretobedifferent18** [comment]
@iamareebjamal I have Improved the code for single click on same page and it looks more cleaner then before

**Comment by daretobedifferent18** [comment]
@iamareebjamal I was reading about ember run loops which I think might help in this case

**Comment by iamareebjamal** [comment]
They help when you want to run some code in next event loop iteration

**Comment by daretobedifferent18** [comment]
@iamareebjamal Is there a way in which I could run a code after the DOM is completely rendered.
didElementInsert only happens first time, I want everytime.

**Comment by iamareebjamal** [comment]
What's the usecase? `MutationObserver` may help but there might be better solution for your specific usecase

**Comment by iamareebjamal** [comment]
After latest changes, I can't go from other page to info link. Go to speakers and then click on sponsors. Not working

**Comment by daretobedifferent18** [comment]
> What's the usecase? `MutationObserver` may help but there might be better solution for your specific usecase

I was thinking of the way in which we can get previous visted route and if the route is speaker or coc etc. we can do dom manipulation like here scrolling. 
Now the problem is I can get previous route but need to do DOM after render is complete.

**Comment by daretobedifferent18** [comment]
> After latest changes, I can't go from other page to info link. Go to speakers and then click on sponsors. Not working

Its working for me

**Comment by iamareebjamal** [comment]
Easier think would be to abandon LinkTo or href-to helper and just use pure anchor tags and see if things work

**Comment by daretobedifferent18** [comment]
> Easier think would be to abandon LinkTo or href-to helper and just use pure anchor tags and see if things work

no it does not work

**Comment by daretobedifferent18** [comment]
@iamareebjamal after lot of things doing and researching for this solution, I found a workaround which will be working very good for TODO-2

**Comment by daretobedifferent18** [comment]
@iamareebjamal I have implemented this method for going to section from diff page.
Pls let me know to do any changes.

**Comment by iamareebjamal** [comment]
It's not going to the section from different page

https://open-event-frontend-git-double-click-5123.eventyay.vercel.app/e/3dbaaa50/coc

Now click on sponsors or organizer

**Comment by iamareebjamal** [comment]
There's no need for goTo action. This could be simplified even more and thus I have done that. Please check if everything is working as expected

**Comment by daretobedifferent18** [comment]
> There's no need for goTo action. This could be simplified even more and thus I have done that. Please check if everything is working as expected

Yep everything is working fine. Thanks for improving it.

**Comment by iamareebjamal** [comment]
Great. Please check gitter

**Pull request merged**